amaranth/vendor/_gatemate.py

```diff
@@ -0,0 +1,120 @@
+from abc import abstractmethod
+from amaranth import *
+from amaranth.build import *
+from amaranth.lib.cdc import ResetSynchronizer
+
+__all__ = ["GateMatePlatform"]
+
+class GateMatePlatform(TemplatedPlatform):
+    """
+    .. rubric:: GateMate FPGA toolchain
+
+    Required tools:
+        * ``yosys``
+        * ``p_r``
+
+    The environment is populated by setting the ``CC_TOOL`` environment variable to point to the toolchain directory.
+    """
+
+    device = property(abstractmethod(lambda: None))
+    package = property(abstractmethod(lambda: None))
+
+    toolchain = "GateMate"
+
+    required_tools = [
+        "yosys",
+        "p_r",
+    ]
+
+    file_templates = {
+        **TemplatedPlatform.build_script_templates,
+        "{{name}}.v": r"""
+            /* {{autogenerated}} */
+            {{emit_verilog()}}
+        """,
+        "{{name}}.debug.v": r"""
+            /* {{autogenerated}} */
+            {{emit_debug_verilog()}}
+        """,
+        "{{name}}.ccf": r"""
+            # {{autogenerated}}
+            {% for port_name, pin_name, attrs in platform.iter_port_constraints_bits() -%}
+                Net "{{port_name}}" Loc = "{{pin_name}}"
+                {%- set constraints = [
+                    "SCHMITT_TRIGGER", "PULLUP", "PULLDOWN", "KEEPER", "SLEW", "DRIVE", 
+                    "DELAY_OBF", "DELAY_IBF", "FF_IBF", "FF_OBF", "LVDS_BOOST", "LVDS_TERM"
+                ] -%}
+                {%- for constraint in constraints if constraint in attrs -%}
+                    | {{constraint}}={{attrs[constraint]}}
+                {%- endfor -%};
+            {% endfor %}
+        """,
+        "{{name}}.sdc": r"""
+            # {{autogenerated}}
+            {% for net_signal, port_signal, frequency in platform.iter_clock_constraints() -%}
```

Comment:
There is no `iter_clock_constraints` function in the `main` branch (you can grep it with zero hits), how did you test `sdc` generation functionality?

Reply:
I didn't test the .sdc function, to be honest. I based this part of the on the `_quicklogic.py` vendor.

What do I need to do here?

Reply:
You need to make sure that a clock constraint, as specified in your board file at least, is actually applied. Whether a clock constraint is applied or not can be seen in the report of the PNR tool, in a format described in the vendor's documentation.

Reply:
I searched through the documentation and couldn't find a clock constraint file anywhere.
This is the workflow.
![image](https://github.com/user-attachments/assets/61808f2d-df1c-4282-bd4b-3417b92c8704)

I will ask a question on our github about the clock constraints

Reply:
That has a clock constraint file in the top right, as I would expect.

Reply:
I think that is their pin constraint file, the `.ccf` file?